text/0000-bom-in-layer-metadata.md

```diff
@@ -0,0 +1,108 @@
+
+This leads to surprising behaviors for users at times when they may be generating a `BOM` as part of the layer generation process and may find that their `BOM` is missing if the `BOM` generation process is not called when the layer is assumed to be cached.
+
+Having the `BOM` be a part of the layer content metadata also means that we can pin-point which layer contributed a BOM entry without having to implement extra logic in downstream buildpacks.
+
+- What is the expected outcome?
+
+The lifecycle is able to generate the final `BOM` by combining the `BOM` entries from top level metadata files like `launch.toml` and `build.toml` and also the ones from `<layer-name>.toml` which handling layer caching correctly i.e. if the layer is cached and re-used in the new build - the `BOM` associated with that layer will also be re-used.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The layer content metadata files would be updated to include - 
+
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
+# arbitrary metadata describing the dependency
+```
+
+For each dependency contributed to the app image or build environment, the buildpack layer:
+
+- SHOULD add a bill-of-materials entry to the bom array describing the dependency, where:
+  - name is REQUIRED.
+  - metadata MAY contain additional data describing the dependency.
+  
+The buildpack SHOULD NOT add bom to a layer content metadata file describing the contents that are not contributed by the same layer.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle would be responsible for the generation of the appropriate metadata files and labels given partial `BOM` entries in the `launch.toml`, `build.toml` and `<layer-name>.toml`. The generation of these artifacts given `BOM` entries in `<layer-name>.toml` would be controlled as follows - 
+
+If the layer has `types.launch` set to `true` then all the `bom` entries will be contributed to the app image.
+If the layer has `types.build` set to `true` then all the `bom` entries will be contributed to the build environment.
+If the layer has `types.cache` set to `true` and if the layer is reused in the next build, the respective `bom` entries should be re-populated from the cached layer content metadata file. If `types.cache` is set to `false` all the `bom` entries from the previous build must be cleaned before/during the export.
+
+All `bom` entries should respect any combination of the above.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+Additional complexity and utility pushed down to the lifecycle.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Keep it as it is and provide a similar functionality in language bindings like `libcnb`. Currently the layer content metadata files are already being used to store and regenerate the `BOM` in buildpacks from the `paketo` project.
+
+# Prior Art
+[prior-art]: #prior-art
+
+<!-- TODO -->
```

Comment:
@ekcasey can you point to the examples you were talking about in paketo where you store the BOM in layer metadata and then restore it in the next build?

Reply:
The paketo Java buildpacks have some related prior art but not the exact thing I think you are asking for. The paketo Java buildpacks regenerate the BOM everytime even if the layer is reused.

prior art:

1. `libpak` also has logic to autogenerate BOM entries for certain types of layers, and it uses the layer types to determine whether it needs a launch or build BOM entry - https://github.com/paketo-buildpacks/libpak/blob/a7abf79a2a53ada27c56d48db765454405ab8259/layer.go#L139-L148
1. We include the layer name in the BOM metadata.